regression-test-harness/proposals/regression_test_harness_upgrades.md

```diff
@@ -0,0 +1,152 @@
+to introduce a much better separation of concerns which is currently lacking.
+
+## Improve flexibility
+
+Once `eosio-launcher` is replaced and reliable process control is centralized
+in the Python code, the next step is to improve the flexibility of the
+test harness by removing all hardcoded assumptions about filesystem
+paths and network ports.  Much of this comes as a side effect of replacing
+`eosio-launcher`.  That replacement work is not simply a line-by-line
+transcription of the existing C++ into similar Python.  Instead, the
+functionality of `eosio-launcher` will be broken up into cluster and node
+control modules, concepts which currently exist in the Python code but
+which are much abbreviated and, in the case of node control, badly stunted.
+
+After the migration, the Python node will be responsible for the filesystem
+paths and network ports used by the `nodeos` instance it is managing.  This
+allows for automated and coordinated collision avoidance.
+
+### Filesystem usage
+
+When a new `nodeos` instance is launched for a test, it writes a substantial
+number of files to disk, including the block log.  These files must be kept
+separate for each instance.  The current hardcoded paths, as mediated by
+`eosio-launcher`, follow a naming scheme intended to be developer-friendly.
+The first node runs out of the `var/lib/node_bios` directory.  Subsequent
+nodes, should they be required, run out of `var/lib/node_00`, 
+`var/lib/node_01`, etc.  This naming scheme allows the developer to make use
+of the test harness during development work.  The log files for each node
+are in predictable, easily remembered locations, making it easy for a developer
+to investigate the causes of a failed test.
+
+This naming scheme will be preserved and extended.  The `var/lib` prefix will
+be removed.  In its place will be a parent directory named after the name of
+the test being run as it is known to `ctest`.  This allows all tests to run
+simultaneously while maintaining the required filesystem isolation without
+the loss of human navigability that would be incurred by randomly generated
+temporary directory names.
+
+In addition, the path names for a cluster and for individual nodes will be
+customizable to facilitate cases where greater clarity can be achieved by
+specific node directory names.
+
+### Network usage
+
+A running `nodeos` instance listens on up to four different network ports,
+depending on which plugins are enabled, and there is a community requested
+feature which would expand that to five.  Berkley sockets restrict listening
+to one process per IP address and port, requiring coordination between nodes
+in a cluster and across clusters to avoid collisions.
+
+The current test harness hardcodes an initial listen port of 8888 on 127.0.0.1
+and uses each port in sequence per node in the cluster for the http plugin.
+The revised system will take advantage of the entirety of the 127.0.0.0/8
```

Comment:
This many not be portable. I don't think macOS routes all of 127.0/8 by default, for example. How difficult will it be to have a fallback for such a platform that still executes sequentially?

Reply:
I wasn't aware macOS is not RFC 3330 compliant.  macOS is not a supported platform.  If and when we support it, this: `sudo ifconfig lo0 alias 127.0.0.2 up` can be added to the test harness when it's running on macOS.  The Python `platform` module can be used to identify macOS as needed.

Reply:
So are you saying for macos you can use a strategy that maps all IP & ports to 127.0.0.2:x?

Reply:
Sorry, I wasn't clear.  Variations of that sudo would be used for all IPs needed in the subnet.  127.0.0.x.

Reply:
What will happen when a user runs `ctest` in this case?

Reply:
When a macOS user runs `ctest` they will get one password prompt on the command line the first time any test runs which requires the loopback addresses.  All tests will check for their presence and silently use them if they're available, which will avoid additional password prompts that might otherwise arise because the sudo timeout has been reached before the entire suite of tests has completed.